text/0070-disallow-empty-namespace-shadowing.md

```diff
@@ -0,0 +1,155 @@
+# Disallow shadowing definitions in the empty namespace
+
+## Related issues and PRs
+
+- Reference Issues: Successor to [RFC 64]
+- Implementation PR(s):
+
+## Timeline
+
+- Started: 2024-06-26
+
+## Summary
+
+In schemas, disallow definitions of entity and common types that would shadow
+definitions of other entity/common types in the empty namespace.
+
+## Basic example
+
+Borrowing and slightly tweaking the example from [RFC 64]:
+
+```
+entity User {
+    email: String,
+};
+
+namespace Demo {
+    entity User {
+        id: String,
+    };
+
+    entity Account {
+        owner: User,
+    };
+}
+```
+
+Today, this schema is accepted, and `Account.owner` is assumed to be
+`Demo::User` and not the empty-namespace `User`.
+As noted in [RFC 64], there is no way for the user to indicate that they intend
+`Account.owner` to be the empty-namespace `User`.
+Rather than add syntax to enable this (as proposed in [RFC 64]), this RFC
+proposes to make this situation an error -- disallowing declaring a `User` type
+in the `Demo` namespace when one already exists in the empty namespace.
+
+## Motivation
+
+Name shadowing is complicated and confusing; schemas using name shadowing may
+look like they are doing one thing when they are actually doing another.
+For Cedar policy and schema authors, this can cause misunderstandings and make
+it more difficult to write correct schemas and policies.
+By disallowing shadowing in the most egregiously ambiguous cases -- when the
+name being shadowed was declared in the empty namespace -- we nudge users to
+make the schema clearer, for instance by renaming a type, or moving a
+conflicting empty-namespace declaration into a new namespace.
+
+## Detailed design
+
+We disallow shadowing a type in the empty namespace with a new declaration in a
+nonempty namespace.
+This applies regardless of whether the shadowed declaration is an entity or
+common type, and whether the shadowing declaration is an entity or common type.
```

Comment:
Would this apply in JSON schema as well? There's never any ambiguity between entity type and and common types in the JSON format, so they're not technically shadowing. Probably worth it to restrict the JSON for the sake of consistency, but the main motivation for this RFC doesn't apply.

Reply:
This is a good point.  I can see arguments both ways.

For the current proposal:
* We probably want to avoid introducing more edge cases where converting JSON to human-syntax schema is difficult or fallible.  (Existing edge cases of this sort are discussed in [RFC 24](https://github.com/cedar-policy/rfcs/blob/main/text/0024-schema-syntax.md).)  
* This RFC is already technically a breaking change (although the break only applies in the particular edge case where users were relying on shadowing an empty-namespace declaration).  Making it "a little more" breaking by disallowing a few more edge-case JSON schemas probably doesn't make much of a difference to end-users.
* We end up in a more consistent state, where more similar rules apply to the human and JSON formats.

For amending the current proposal, to allow an entity type with the same name as an empty-namespace common type, and vice versa, in the JSON format only:
* You are right that there's no technical reason we can't allow this; these declarations do not conflict and are not ambiguous in the JSON format, which distinguishes entity typenames from common typenames.
* It makes this RFC a little less breaking, by avoiding breaking users who were already doing this in a JSON schema.

Reply:
We have a rule (guideline?) that you can convert back-and-forth between JSON and human readable formats. If you allow this in JSON but not human-readable we break the tenet. IMO we should change the JSON.

Reply:
Based on this thread and @khieta's review below, I will amend the RFC to make clear we will disallow this in the JSON as well.

In particular I'm referencing these comments:

@john-h-kastner-aws: "Probably worth it to restrict the JSON for the sake of consistency"
@cdisselkoen: "We end up in a more consistent state, where more similar rules apply to the human and JSON formats"
@mwhicks1: "If you allow this in JSON but not human-readable we break the tenet. IMO we should change the JSON"
@khieta: "My opinion is that we should enforce the same restrictions on both"

Reply:
Added text at the end of the `Detailed design` section